Add stats to OutputBuffer

HolyLow opened the pull request:
Add runtime stats for OutputBuffer: is-finished flag, number of bytes/rows/pages buffered and sent.

These stats will be used in Prestissimo to populate TaskStats fields.

For each destination buffer, the stats are:
    - finished: If this destination buffer is finished or not. Set when the destination buffer is called `finish()`.
    - bytesBuffered, rowsBuffered, pagesBuffered: the number of bytes/rows/pages buffered in the destination buffer to be consumed;
    - bytesSent, rowsSent, pagesSent: the number of bytes/rows/pages sent from the destination buffer and acked via acknowledge() or deleted via deleteResults().

For each output buffer, the stats are:
    - kind: partitioned, broadcast or arbitrary
    - noMoreBuffers: indicates whether the output buffer could accept more destination buffers (allowed for broadcast and arbitrary only)
    - noMoreData: indicates whether the output buffer has received all data from the producers
    - finished: indicates whether the output buffer has finished all data transferring
    - bufferStats: stats for individual destination buffers

HolyLow commented:
@mbasmanova @kewang1024 Would you kindly review this PR please? It seems that you are the main contributors of the OutputBuffer related logic.

mbasmanova reviewed (commented):
@HolyLow Thank you for adding stats for OutputBuffer. Would you update PR description to list stats being added, describe their meaning and how they are collected.

mbasmanova commented on velox/exec/OutputBuffer.h:
```diff
@@ -74,6 +74,41 @@ class ArbitraryBuffer {
 
+struct BufferInfo {
+  void recordAddPage(const SerializedPage& data) {
+    VELOX_CHECK(data.rows() >= 0, "SerializedPage's row size is not inited");
```
VELOX_CHECK_GE for better error messages

HolyLow replied:
done.

mbasmanova commented on velox/exec/OutputBuffer.h:
```diff
@@ -74,6 +74,41 @@ class ArbitraryBuffer {
+    rowsAdded += rows;
+    pagesAdded++;
+  }
+  void recordSendPage(const SerializedPage& data) {
```
add empty line

HolyLow replied:
done.

mbasmanova commented on velox/exec/OutputBuffer.h:
```diff
@@ -74,6 +74,41 @@ class ArbitraryBuffer {
+    rowsSent += rows;
+    pagesSent++;
+  }
+  bool finished{false};
```
add empty line

Add comments to each field.

HolyLow replied:
done.

mbasmanova commented on velox/exec/OutputBuffer.h:
```diff
@@ -74,6 +74,41 @@ class ArbitraryBuffer {
+  int64_t bufferedBytes{0};
+  int64_t bufferedRows{0};
+  int64_t bufferedPages{0};
+  int64_t bytesAdded{0};
```
consistency: bufferedBytes vs. addedBytes

maybe: bytesBuffered, bytesAdded, bytesSent

HolyLow replied:
done.

mbasmanova commented on velox/exec/OutputBuffer.h:
```diff
@@ -114,6 +149,8 @@ class DestinationBuffer {
   // the callback.
   DataAvailable getAndClearNotify();
 
+  BufferInfo getInfo();
```
Maybe BufferStats?

getInfo() -> stats()

Make this method const

lingbin replied:
+1 for `BufferStats`.

HolyLow replied:
done.

mbasmanova commented on velox/exec/OutputBuffer.h:
```diff
@@ -124,10 +161,36 @@ class DestinationBuffer {
 
+struct OutputBufferInfo {
+  enum class BufferState {
+    OPEN,
```
naming: kOpen

please, check coding guidelines and update

HolyLow replied:
done.

mbasmanova commented on velox/exec/tests/OutputBufferManagerTest.cpp:
```diff
@@ -960,6 +964,41 @@ TEST_P(AllOutputBufferManagerTest, outputBufferUtilization) {
+    totalSize += size;
+    // force Arbitrary buffer to load data.
+    if (kind_ ==
+        facebook::velox::core::PartitionedOutputNode::Kind::kArbitrary) {
```
drop facebook::velox::

HolyLow replied:
done.

mbasmanova reviewed (commented):
Some initial comments.

lingbin commented on velox/exec/OutputBuffer.h:
```diff
@@ -124,10 +161,36 @@ class DestinationBuffer {
 
 class Task;
 
+struct OutputBufferInfo {
```
Please add some comments for this class, and its relationship with `BufferInfo`, whose names are very similar.  
 
Many Velox users are not Presto users, so they are not familiar with these names and may be confused.

HolyLow replied:
done.

lingbin commented on velox/exec/OutputBuffer.h:
```diff
@@ -124,10 +161,36 @@ class DestinationBuffer {
+  int64_t totalPagesAdded{0};
+  int64_t totalBytesSent{0};
+  int64_t totalRowsSent{0};
+  int64_t totalPagesSent{0};
```
Let's keep naming consistent: `totalBufferedBytes`-> `totalBytesBuffered`?

HolyLow replied:
done.

lingbin commented on velox/exec/ExchangeQueue.h:
```diff
@@ -26,7 +26,8 @@ class SerializedPage {
       std::unique_ptr<folly::IOBuf> iobuf,
-      std::function<void(folly::IOBuf&)> onDestructionCb = nullptr);
+      std::function<void(folly::IOBuf&)> onDestructionCb = nullptr,
+      int64_t rows = -1);
```
nit: If the default value of `rows` is `0`, then we no longer need a lot of `VELOX_CHECK(data.rows() >= 0)` when getting its row count, right?

HolyLow replied:
Nop. The `row` param is not always provided, because for `ExchangeSource` the rows of a serializedPage is not known until the page gets unserialized. So the `row` param should be optional, and `-1` is necessary to indicates if the `row` param is valid or not.

lingbin commented on velox/exec/ExchangeQueue.h:
```diff
@@ -35,6 +36,12 @@ class SerializedPage {
     return iobufBytes_;
   }
 
+  // Returns the rows of the serialized data as a page.
```
nit: "Returns the rows in current serialized page"?

lingbin commented on velox/exec/ExchangeQueue.h:
```diff
@@ -61,6 +68,10 @@ class SerializedPage {
   const int64_t iobufBytes_;
 
+  // Number of rows in 'iobuf_' as a page.
+  // -1 means this param is not valid.
```
nit: only keep：
```
// Number of rows in 'iobuf_'.
```

HolyLow replied:
See comments above. -1 is necessary.

lingbin commented on velox/exec/OutputBuffer.cpp:
```diff
@@ -640,4 +654,47 @@ bool OutputBuffer::isOverutilized() const {
+  }
+  if (info.state == OutputBufferInfo::BufferState::FLUSHING &&
+      info.totalBufferedBytes == 0) {
+    info.state = OutputBufferInfo::BufferState::FINISHED;
```
A little confusing: if there is already data, then `totalBufferedBytes` will no longer be reduced to `0`, right? So the `FINISHED` state cannot be reached?

HolyLow replied:
Please read the newly added annotations for `xxxbufferedBytes`.

lingbin commented on velox/exec/OutputBuffer.cpp:
```diff
@@ -640,4 +654,47 @@ bool OutputBuffer::isOverutilized() const {
+    info.totalBytesSent += bufferInfo.bytesSent;
+    info.totalRowsSent += bufferInfo.rowsSent;
+    info.totalPagesSent += bufferInfo.pagesSent;
+  }
```
Does the update of these members (eg, `totalBufferedBytes` / `totalBufferedRows`) also need to be carried out under the protection of `mutex_`?

HolyLow replied:
No, they don't.

HolyLow commented on velox/exec/OutputBufferManager.cpp:
```diff
@@ -183,4 +183,9 @@ bool OutputBufferManager::isOverutilized(const std::string& taskId) {
 }
 
+OutputBufferInfo OutputBufferManager::getInfo(const std::string& taskId) {
+  auto buffer = getBuffer(taskId);
```
No, `getBuffer()` would never return `nullptr`. `getBufferIfExists()` would return `nullptr`, but as we are getting the stats of a task's outputBuffer, the outputBuffer should exist.

lingbin replied:
Thank you for your explanation, `getBuffer()` does not return `nullptr`, but it may throw an exception.

No offense, I just wonder about the better behavior of `getStats()` if this "task" no longer exists (this must happen, right? That is, after the buffer ends, another thread calls the function)?

If an exception is thrown, the caller needs additional `try...catch...`.

At the same time, I noticed that another API for obtaining information (`getUtilization()`) will face the same situation, but it will not throw an exception.

HolyLow replied:
Many existing interfaces including `updateNumDrivers()`, `enqueue()`, `noMoreData()`, `isFinished()` use the `getBuffer()` function to get the buffer without checking. For these interfaces, the buffer **should** exist, otherwise an exception is expected. I think this is by design. 

If you have doubts on the existing `OutputBufferManager`'s design, maybe you should open another issue to discuss it rather than discuss here. This PR is not for it.

HolyLow commented on velox/exec/OutputBuffer.h:
```diff
@@ -74,6 +74,41 @@ class ArbitraryBuffer {
   std::deque<std::shared_ptr<SerializedPage>> pages_;
 };
 
+struct BufferInfo {
```
done.

HolyLow commented on velox/exec/OutputBuffer.h:
```diff
@@ -124,10 +161,36 @@ class DestinationBuffer {
+    FINISHED,
+    FAILED
+  };
+  std::string type;
```
done.

HolyLow commented:
@mbasmanova Thanks for your comments, the PR description is updated, and the code is refactored as suggested. Looking forward to further comments.

mbasmanova commented on velox/exec/OutputBuffer.h:
```diff
@@ -124,10 +184,58 @@ class DestinationBuffer {
+  /// The bufferStats of each destination buffer.
+  std::vector<BufferStats> buffers;
+  /// Total byte num of data buffered in all destination buffers.
+  int64_t totalBytesBuffered{0};
```
It is the case that totalAdded = totalBuffered + totalSent? If this is so, then perhaps, we can remove totalAdded as these are redundant.

HolyLow replied:
Yes, totalAdded = totalBuffered + totalSent, but I tend to keep the "redundant" totalAdded, because the 'added' -> 'buffered' -> 'sent' are three different phases for an enqueued page, and keep each phase recorded separately is **more straightforward and easier to understand**.

Well, if you insist to remove the redundancy, I could remove the field then. Also note that actually all the `totalXXX` fields in `OutputBufferStats` are simply accumulations of corresponding `BufferStats`, which could be considered as redundant as well. So if we are going to reduce all the redundancy, all the `totalXXX` fields could be removed.

mbasmanova replied:
@HolyLow These are good points. It would be nice to reduce redundancy. Thanks.

HolyLow replied:
done.

mbasmanova commented on velox/exec/OutputBuffer.cpp:
```diff
@@ -640,4 +654,46 @@ bool OutputBuffer::isOverutilized() const {
+  OutputBufferStats stats;
+  {
+    std::lock_guard<std::mutex> l(mutex_);
+    stats.type = PartitionedOutputNode::kindString(kind_);
```
Can we use the kind_ enum here? Why do we need to convert it into a string?

HolyLow replied:
I was thinking that the string type here would be more easy to understand. But you've made a point here, and I change the string type to the kind enum in latest update.

mbasmanova commented on velox/exec/OutputBuffer.cpp:
```diff
@@ -640,4 +654,46 @@ bool OutputBuffer::isOverutilized() const {
+  {
+    std::lock_guard<std::mutex> l(mutex_);
+    stats.type = PartitionedOutputNode::kindString(kind_);
+    stats.state = OutputBufferStats::BufferState::kOpen;
```
It is not a good practice to set a variable to "wrong" value, then fix it up. Let's create a helper method to return the buffer state.

```
if (isFinishedLocked()) {
   return kFinished;
}

if (atEnd_) {
   return kFlushing;
}

...
```

HolyLow replied:
Done. Code moved to `OutputBufferStats`'s constructor with the form you propose.

mbasmanova commented on velox/exec/OutputBuffer.h:
```diff
@@ -124,10 +184,58 @@ class DestinationBuffer {
+    /// The OutputBuffer has already received and sent all the input data.
+    kFinished,
+    /// The OutputBuffer has failed.
+    kFailed
```
I don't see this being used. Is it?

HolyLow replied:
Yes, it is not used yet. Removed.

mbasmanova reviewed (commented):
It seems like this PR includes many changes. Would it be possible to break it up into smaller pieces? This would make it easier to review.

mbasmanova commented on velox/exec/ExchangeQueue.h:
```diff
@@ -26,7 +26,8 @@ class SerializedPage {
       std::unique_ptr<folly::IOBuf> iobuf,
-      std::function<void(folly::IOBuf&)> onDestructionCb = nullptr);
+      std::function<void(folly::IOBuf&)> onDestructionCb = nullptr,
+      int64_t rows = -1);
```
@HolyLow Would it be possible to extract this change into a separate PR and use std::optional instead of magic value?

HolyLow replied:
This change is extracted to another PR: https://github.com/facebookincubator/velox/pull/8043 . Please help review it, thanks a lot.

HolyLow commented:
@mbasmanova the code has reduced redundancy and rebased to the SerializedPage's modification. Looking forward to further comments,  thanks.

mbasmanova commented on velox/exec/OutputBuffer.h:
```diff
@@ -74,6 +74,36 @@ class ArbitraryBuffer {
   std::deque<std::shared_ptr<SerializedPage>> pages_;
 };
 
+/// The stats of a DestinationBuffer. The data transferred by the buffer has
```
There is a bit of inconsistency in naming throughout this change. Multiple different names are used for the same thing. Many comments contain repetitive / redundant text. Let's fix that. I'll point out a few examples and hopefully it will be clear what I mean. 

BufferStats -> DestinationBufferStats

> The stats of a DestinationBuffer.

This statement is redundant. It just repeats the struct name.

HolyLow replied:
done.

mbasmanova commented on velox/exec/OutputBuffer.h:
```diff
@@ -74,6 +74,36 @@ class ArbitraryBuffer {
 
+/// The stats of a DestinationBuffer. The data transferred by the buffer has
+/// two phases:
+/// 1. Buffered: after added via enqueue() and before acked, the data is
```
Notice, the use of 2 terms, add and enqueue, for the same thing. Let's use "enqueue".

recordEnqueue

HolyLow replied:
done.

mbasmanova commented on velox/exec/OutputBuffer.h:
```diff
@@ -74,6 +74,36 @@ class ArbitraryBuffer {
+/// 2. Sent: the data is sent and removed from the buffer after it is acked
+///          via acknowledge() or deleteResults().
+struct BufferStats {
+  /// Update stats when a data page is added.
```
This comment is redundant.

HolyLow replied:
removed.

mbasmanova commented on velox/exec/OutputBuffer.h:
```diff
@@ -74,6 +74,36 @@ class ArbitraryBuffer {
+  void recordAddPage(const SerializedPage& data);
+
+  /// Update stats when a data page is sent (removed from buffer).
+  void recordSendPage(const SerializedPage& data);
```
recordAcknowledge + recordDelete (the implementation of these two will be the same, but the API will be clearer)

HolyLow replied:
done.

mbasmanova commented on velox/exec/OutputBuffer.h:
```diff
@@ -74,6 +74,36 @@ class ArbitraryBuffer {
+  /// Update stats when a data page is sent (removed from buffer).
+  void recordSendPage(const SerializedPage& data);
+
+  /// If the destination buffer is finished. A finished dest buffer would be
```
redundant comment; let's explain what "finished" mean instead.

HolyLow replied:
done.

mbasmanova commented on velox/exec/OutputBuffer.h:
```diff
@@ -74,6 +74,36 @@ class ArbitraryBuffer {
+  /// If the destination buffer is finished. A finished dest buffer would be
+  /// deleted by the OutputBuffer, so this flag is set by the OutputBuffer.
+  bool finished{false};
+  /// Byte num of data that is currently buffered.
```
Separate properties with empty lines for readability.

Comments are redundant as they repeat the names of the variables. The explanation of buffered vs. sent in the struct-level comment is sufficient.

```
int64_t bytesBuffered{0};
int64_t rowsBuffered{0};
int64_t pagesBuffered{0};

int64_t bytesSent{0};
int64_t rowsSent{0};
int64_t pagesSent{0};
```

HolyLow replied:
done.

mbasmanova commented on velox/exec/OutputBuffer.h:
```diff
@@ -114,6 +144,9 @@ class DestinationBuffer {
   DataAvailable getAndClearNotify();
 
+  // Returns the stats of this buffer.
+  BufferStats getStats() const;
```
remove 'get' prefix

DestinationBufferStats stats() const;

Consider moving DestinationBufferStats into DestinationBuffer class: DestinationBuffer::Stats

HolyLow replied:
`get` prefix removed, but after some trying I tend to leave DestinationBufferStats outside of DestinationBuffer because moving it inside would make the DestinationBuffer class harder to read.

mbasmanova commented on velox/exec/OutputBuffer.h:
```diff
@@ -124,10 +157,45 @@ class DestinationBuffer {
 
 class Task;
 
+/// The stats of an OutputBuffer.
```
redundant comment

HolyLow replied:
removed.

mbasmanova commented on velox/exec/OutputBuffer.h:
```diff
@@ -124,10 +157,45 @@ class DestinationBuffer {
 
+/// The stats of an OutputBuffer.
+struct OutputBufferStats {
+  /// BufferState indicates the OutputBuffer's current state.
```
redundant comment

HolyLow replied:
removed.

mbasmanova commented on velox/exec/OutputBuffer.h:
```diff
@@ -124,10 +157,45 @@ class DestinationBuffer {
+struct OutputBufferStats {
+  /// BufferState indicates the OutputBuffer's current state.
+  enum class BufferState {
+    /// The OutputBuffer is open and ready to receive data. Initial state.
```
/// Initial state.

mbasmanova commented on velox/exec/OutputBuffer.h:
```diff
@@ -124,10 +157,45 @@ class DestinationBuffer {
+  enum class BufferState {
+    /// The OutputBuffer is open and ready to receive data. Initial state.
+    kOpen,
+    /// The OutputBuffer would not accept more buffers.
```
/// Doesn't accept more buffers. Partitioned buffers specify number of destinations at creation and do not allow adding buffers. Broadcast and Arbitrary buffers allow adding buffers.

HolyLow replied:
removed.

mbasmanova commented on velox/exec/OutputBuffer.h:
```diff
@@ -124,10 +157,45 @@ class DestinationBuffer {
+    kOpen,
+    /// The OutputBuffer would not accept more buffers.
+    kNoMoreBuffers,
+    /// The OutputBuffer has received all input data, and some data is still
```
Received all data. Waiting for the data to be consumed.

HolyLow replied:
removed.

mbasmanova commented on velox/exec/OutputBuffer.h:
```diff
@@ -124,10 +157,45 @@ class DestinationBuffer {
+    /// The OutputBuffer has received all input data, and some data is still
+    /// to be sent.
+    kFlushing,
+    /// The OutputBuffer has already received and sent all the input data.
```
All data has been received and consumed.

HolyLow replied:
removed.

mbasmanova commented on velox/exec/OutputBuffer.h:
```diff
@@ -124,10 +157,45 @@ class DestinationBuffer {
+
+  OutputBufferStats(
+      core::PartitionedOutputNode::Kind k,
+      bool noMoreBuffers,
```
noMoreBuffers vs. canAddBuffers

atEnd vs. canAddPages

Do we need canAddBuffers and canAddPages? Is state not sufficient?

HolyLow replied:
After some consideration, I decide to remove the `BufferState` and keep these flags.

 I think the state flags including `canAddBuffers`, `canAddPages` and `finished` are more descriptive than the previous `BufferState` as these they would describe every aspects of the OutputBuffer's status rather than fusing the states into a unified BufferState.
Besides, different engines like presto, spark might have different BufferState definitions, and they could use the state flags to translate to their own version of BufferState.

mbasmanova commented on velox/exec/OutputBuffer.h:
```diff
@@ -124,10 +157,45 @@ class DestinationBuffer {
+      const std::vector<BufferStats>& bufs);
+
+  /// The kind of the OutputBuffer.
+  core::PartitionedOutputNode::Kind kind;
```
make all of these const
revise comment to remove redundant text

HolyLow replied:
done.

mbasmanova commented on velox/exec/OutputBuffer.h:
```diff
@@ -124,10 +157,45 @@ class DestinationBuffer {
+  /// If the outputBuffer can receive more input pages.
+  bool canAddPages{true};
+  /// The bufferStats of each destination buffer.
+  std::vector<BufferStats> buffers;
```
destinationBuffers

HolyLow replied:
done.

mbasmanova commented on velox/exec/OutputBuffer.h:
```diff
@@ -192,6 +260,9 @@ class OutputBuffer {
   bool isOverutilized() const;
 
+  // Gets the OutputBufferStats of this output buffer.
+  OutputBufferStats getStats();
```
stats() const

HolyLow replied:
Refactored to `stats()`, but the `stats()` function here could not be const because the mutex is used in lock guard.

mbasmanova commented on velox/exec/OutputBuffer.h:
```diff
@@ -272,6 +343,8 @@ class OutputBuffer {
   int32_t nextArbitraryLoadBufferIndex_{0};
   // One buffer per destination.
   std::vector<std::unique_ptr<DestinationBuffer>> buffers_;
+  // One bufferStat per destination.
```
Redundant comment. It would be helpful to explain why we need this member variable and cannot just use buffers_[i].stats(). I assume this variable caches final stats for deleted buffers. Is this right? The number of entries in buffers_ and bufferStats_ is the same, but bufferStats_[i] is set only if buffers_[i] is null and vice versa.

Perhaps, finishedBufferStats_ would be a more descriptive name.

HolyLow replied:
done.

mbasmanova commented on velox/exec/OutputBuffer.cpp:
```diff
@@ -63,6 +63,29 @@ std::string ArbitraryBuffer::toString() const {
 }
 
+void BufferStats::recordAddPage(const SerializedPage& data) {
+  auto numRows = data.numRows();
```
const

HolyLow replied:
done.

mbasmanova commented on velox/exec/OutputBuffer.cpp:
```diff
@@ -63,6 +63,29 @@ std::string ArbitraryBuffer::toString() const {
 
+void BufferStats::recordAddPage(const SerializedPage& data) {
+  auto numRows = data.numRows();
+  VELOX_CHECK(numRows, "SerializedPage's numRows must be valid");
```
VELOX_CHECK(numRows.has_value()...)

HolyLow replied:
`has_value()` is indeed more descriptive. refactored.

mbasmanova commented on velox/exec/OutputBuffer.cpp:
```diff
@@ -63,6 +63,29 @@ std::string ArbitraryBuffer::toString() const {
+void BufferStats::recordAddPage(const SerializedPage& data) {
+  auto numRows = data.numRows();
+  VELOX_CHECK(numRows, "SerializedPage's numRows must be valid");
+  int64_t rows = numRows.value();
```
remove these temp variables; not needed

HolyLow replied:
done.

mbasmanova commented on velox/exec/OutputBufferManager.h:
```diff
@@ -108,6 +108,9 @@ class OutputBufferManager {
   bool isOverutilized(const std::string& taskId);
 
+  // Gets the OutputBufferStats of the taskId's outputBuffer.
+  OutputBufferStats getStats(const std::string& taskId);
```
stats

HolyLow replied:
done.

mbasmanova commented on velox/exec/OutputBufferManager.h:
```diff
@@ -108,6 +108,9 @@ class OutputBufferManager {
   // producers. When the task of this taskId is not found, return false.
   bool isOverutilized(const std::string& taskId);
 
+  // Gets the OutputBufferStats of the taskId's outputBuffer.
```
Redundant comments. Let's explain what happens if task no longer exists.

HolyLow replied:
done.

mbasmanova commented on velox/exec/OutputBufferManager.cpp:
```diff
@@ -183,4 +183,9 @@ bool OutputBufferManager::isOverutilized(const std::string& taskId) {
 }
 
+OutputBufferStats OutputBufferManager::getStats(const std::string& taskId) {
+  auto buffer = getBuffer(taskId);
```
I think this API should use getBufferIfExists and handle task missing or already finished and removed cases gracefully. Maybe it can return std::optional<OutputBufferStats>

HolyLow replied:
done.

mbasmanova reviewed (commented):
@HolyLow Thank you for iterating on this PR. It is shaping up nicely. These stats will be very useful. Some comments. Haven't reviewed the test code yet. Will do so once current comments are taken care of. Thanks.

mbasmanova commented:
@HolyLow Please, update PR description to match the code.

HolyLow commented:
> @HolyLow Please, update PR description to match the code.

@mbasmanova The PR description is updated, and the code is refactored as requested. Looking forward to further comments, thanks.

mbasmanova commented:
@HolyLow Thank you for replying / addressing comments. I'm going to take a look at the updated code, but I see that CI is red. Would you check it out?

mbasmanova commented on velox/exec/OutputBuffer.h:
```diff
@@ -74,6 +74,33 @@ class ArbitraryBuffer {
+  /// OutputBuffer, so this flag is set by the OutputBuffer.
+  bool finished{false};
+
+  /// Stats of data that is currently buffered.
```
Number of buffered bytes / rows / pages.

HolyLow replied:
done.

HolyLow replied:
Actually your proposed annotation is just like repeating the variables' names, which is not encouraged in our former discussions. 

This PR makes me quite confused about the annotation standard or method in velox. It seems that just mimicking existing annotations in velox would not work (as I did in several iters in this PR) because many existing annotations in velox don't match this PR's standard.

Anyway, I adjust the annotations to what you propose. Thanks for your detailed feedbacks and suggestions.

mbasmanova replied:
@HolyLow This is a good point. I suggested a change because "Stats of data" is too generic and doesn't provide much insight on top of what variable names already provide. I think you have a good point that, perhaps, new comment is similarly redundant. Maybe it be be removed completely. I would keep it though since it helps to summarize the code below in a format that's a bit easier to read than dense code below it. I don't have a strong opinion here though except that we shouldn't use generic "Stats of data".

mbasmanova commented on velox/exec/OutputBuffer.h:
```diff
@@ -74,6 +74,33 @@ class ArbitraryBuffer {
+  int64_t rowsBuffered{0};
+  int64_t pagesBuffered{0};
+
+  /// Stats of data that has been sent from the buffer.
```
Number of sent bytes / rows / pages.

HolyLow replied:
done.

mbasmanova commented on velox/exec/OutputBuffer.h:
```diff
@@ -124,10 +154,35 @@ class DestinationBuffer {
+  const core::PartitionedOutputNode::Kind kind;
+
+  /// States of the OutputBuffer.
+  const bool canAddBuffers;
```
It might be hard to connect these flags to OutputBuffer API. Perhaps, rename to noMoreBuffers and noMoreData. These would match OutputBuffer::noMoreData() API and noMoreBuffers flags passed to OutputBuffer::updateBuffers.

HolyLow replied:
Make sense, modified.

mbasmanova commented on velox/exec/tests/OutputBufferManagerTest.cpp:
```diff
@@ -971,6 +977,60 @@ TEST_P(AllOutputBufferManagerTest, outputBufferUtilization) {
+  const vector_size_t size = 100;
+  const std::string taskId = std::to_string(rand());
+  initializeTask(taskId, rowType_, kind_, 1, 1);
+  OutputBufferStats stats0 = getStats(taskId);
```
variable names a1, a2, a2 are an anti-pattern; Here, stats0 is not needed at all as you can simply do

ASSERT_EQ(getStats(taskId).kind, kind_);

HolyLow replied:
refactored. I kind of `learned` the bad habit of a1, a2, aXX from other projects, will change it. Thanks.

mbasmanova reviewed (commented):
@HolyLow Overall looks good % some nits and new test is hard to follow. Wondering if there is a way to refactor the test to make it easier to read.

HolyLow commented:
> @HolyLow Thank you for replying / addressing comments. I'm going to take a look at the updated code, but I see that CI is red. Would you check it out?

@mbasmanova It seems that the red CI cases are about the memoryPool related logics, which has broken the CI procedures since last week.

The code is modified, looking forward to your further comments. Thanks a lot.

mbasmanova commented:
> It seems that the red CI cases are about the memoryPool related logics, which has broken the CI procedures since last week.

Have you tried rebasing? We won't be able to merge until CI is green. CC: @xiaoxmeng

mbasmanova commented:
> The code is modified,

Somehow I do not see changes. Any chance you forgot to push the latest code?

HolyLow commented:
> > The code is modified,
> 
> Somehow I do not see changes. Any chance you forgot to push the latest code?

The code is pushed, and the commit is https://github.com/facebookincubator/velox/pull/7962/commits/a2c58ebc54dfafa95276d17befd523948b190f4e

HolyLow commented:
> > It seems that the red CI cases are about the memoryPool related logics, which has broken the CI procedures since last week.
> 
> Have you tried rebasing? We won't be able to merge until CI is green. CC: @xiaoxmeng

I tried rebasing several hours before, now I try again to see if CI passes.

mbasmanova commented:
@HolyLow I think test failures might be related to changes in this PR. Here is what I see in CI logs:

https://circleci.com/api/v1.1/project/github/facebookincubator/velox/283004/output/107/0?file=true&allocation-id=658175d88c68da6036352d11-0-build%2FABCDEFGH
```
276: E20231219 11:27:06.955521 63066 Exceptions.h:69] Line: ../../velox/exec/OutputBuffer.cpp:68, Function:recordEnqueue, Expression: numRows.has_value() SerializedPage's numRows must be valid, Source: RUNTIME, ErrorCode: INVALID_STATE
276: unknown file: Failure
276: C++ exception with description "Exception: VeloxRuntimeError
276: Error Source: RUNTIME
276: Error Code: INVALID_STATE
276: Reason: SerializedPage's numRows must be valid
276: Retriable: False
276: Expression: numRows.has_value()
276: Function: recordEnqueue
276: File: ../../velox/exec/OutputBuffer.cpp
276: Line: 68
```

HolyLow commented:
@mbasmanova Fixed the uts, and currently all CI passes except the "Ubuntu Benchmark" CI, whose error log seems not relevant. Should there be more fixes? Or it just doesn't matter...?

mbasmanova commented on velox/exec/tests/OutputBufferManagerTest.cpp:
```diff
@@ -971,6 +977,60 @@ TEST_P(AllOutputBufferManagerTest, outputBufferUtilization) {
+    ASSERT_EQ(statsAck.destinationBuffers[0].rowsBuffered, 0);
+  }
+
+  // set outputBuffer to NoMoreBuffers and check stats
```
nit: Make sure comments are full sentences. Start with a capital letter and end with a period.

HolyLow replied:
quick fixed.

mbasmanova reviewed (commented):
@HolyLow Thank you for fixing the tests. Let me try to merge this PR. 

@kgpai @assignUser  Krishna, Jacob, would you help look into benchmark job failure?

mbasmanova commented:
@HolyLow PR description seems outdated.

> Add runtime stats for OutputBuffer: state, number of bytes added/sent/buffered.

amitkdutta commented on velox/exec/tests/OutputBufferManagerTest.cpp:
```diff
@@ -971,6 +977,60 @@ TEST_P(AllOutputBufferManagerTest, outputBufferUtilization) {
 
+TEST_P(AllOutputBufferManagerTest, outputBufferStats) {
+  const vector_size_t size = 100;
+  const std::string taskId = std::to_string(rand());
```
@HolyLow Meta internal static linting is suggesting following
```
using C rand() to generate random number causes lock contention, please consider folly::Random::rand32().
```
Lets remove rand() here and use [folly](folly::Random::rand32() instead.

HolyLow replied:
done.

xiaoxmeng commented on velox/exec/OutputBuffer.cpp:
```diff
@@ -556,7 +592,10 @@ bool OutputBuffer::deleteResults(int destination) {
     dataAvailable = buffer->getAndClearNotify();
+    VELOX_CHECK_LT(destination, finishedBufferStats_.size());
+    finishedBufferStats_[destination] = buffers_[destination]->stats();
     buffers_[destination] = nullptr;
```
nit: is it better to provide a finish method for destination buffer stats to call finish() before set it to finishedBufferStats_? Thanks!

HolyLow replied:
done.

xiaoxmeng commented on velox/exec/OutputBuffer.cpp:
```diff
@@ -651,4 +690,22 @@ bool OutputBuffer::isOverutilized() const {
+      bufferStats[i] = finishedBufferStats_[i];
+    }
+  }
+  OutputBufferStats stats(
```
```
nit: return OutputBufferStats();
```

HolyLow replied:
done.

xiaoxmeng commented on velox/exec/OutputBufferManager.h:
```diff
@@ -108,6 +108,9 @@ class OutputBufferManager {
   // producers. When the task of this taskId is not found, return false.
   bool isOverutilized(const std::string& taskId);
 
+  // Returns nullopt when the taskId doesn't exist.
```
```
... when the specified output buffer doesn't exist.
```

HolyLow replied:
done.

xiaoxmeng reviewed (approved):
@HolyLow LGTM % nits. Thanks for the improvement!

assignUser commented:
@mbasmanova The benchmark failure is unrelated to the PR it was caused by [main](https://github.com/facebookincubator/velox/actions/runs/7268011818/job/19803048721#step:14:227) but seems to now be fixed. So a rebase should fix it for this PR as well.

HolyLow commented:
@assignUser I rebased the code to main branch, but seems a UT called "velox_cache_test" failed in CI. Don't know what happened.
@xiaoxmeng Refactored the code as requested, looking forward to further comments.

HolyLow commented on velox/exec/OutputBuffer.cpp:
```diff
@@ -63,6 +63,31 @@ std::string ArbitraryBuffer::toString() const {
+  VELOX_CHECK(numRows.has_value(), "SerializedPage's numRows must be valid");
+  const int64_t size = data.size();
+  bytesBuffered -= size;
+  rowsBuffered -= numRows.value();
```
done.

HolyLow commented on velox/exec/OutputBuffer.cpp:
```diff
@@ -63,6 +63,31 @@ std::string ArbitraryBuffer::toString() const {
+  const auto numRows = data.numRows();
+  VELOX_CHECK(numRows.has_value(), "SerializedPage's numRows must be valid");
+  const int64_t size = data.size();
+  bytesBuffered += size;
```
done.

HolyLow commented on velox/exec/OutputBuffer.h:
```diff
@@ -124,10 +154,35 @@ class DestinationBuffer {
 class Task;
 
+struct OutputBufferStats {
+  OutputBufferStats(
```
done.

HolyLow commented on velox/exec/OutputBuffer.h:
```diff
@@ -124,10 +154,35 @@ class DestinationBuffer {
+        noMoreBuffers(noMoreBuffer),
+        noMoreData(noMoreDat),
+        finished(finish),
+        destinationBuffers(buffers) {}
```
done.

HolyLow commented on velox/exec/OutputBuffer.h:
```diff
@@ -124,10 +154,35 @@ class DestinationBuffer {
+      bool noMoreBuffer,
+      bool noMoreDat,
+      bool finish,
+      const std::vector<DestinationBufferStats>& buffers)
```
done.

HolyLow commented on velox/exec/OutputBuffer.h:
```diff
@@ -124,10 +154,35 @@ class DestinationBuffer {
+      core::PartitionedOutputNode::Kind k,
+      bool noMoreBuffer,
+      bool noMoreDat,
+      bool finish,
```
done.

HolyLow commented on velox/exec/OutputBuffer.h:
```diff
@@ -124,10 +154,35 @@ class DestinationBuffer {
+  OutputBufferStats(
+      core::PartitionedOutputNode::Kind k,
+      bool noMoreBuffer,
+      bool noMoreDat,
```
done.

HolyLow commented on velox/exec/OutputBuffer.h:
```diff
@@ -124,10 +154,35 @@ class DestinationBuffer {
 
+struct OutputBufferStats {
+  OutputBufferStats(
+      core::PartitionedOutputNode::Kind k,
```
done.

HolyLow commented on velox/exec/OutputBuffer.h:
```diff
@@ -124,10 +154,35 @@ class DestinationBuffer {
 
 class Task;
 
+struct OutputBufferStats {
```
done.

HolyLow commented on velox/exec/OutputBuffer.h:
```diff
@@ -74,6 +74,33 @@ class ArbitraryBuffer {
+/// 1. Buffered: the data resides in the buffer after enqueued and before
+///              acked / deleted.
+/// 2. Sent: the data is removed from the buffer after it is acked or
+///          deleted.
```
done.

mbasmanova commented:
@HolyLow CI failure is tracked in #8139 CC: @gggrace14

mbasmanova commented:
@HolyLow Curious if you have a Prestissimo PR to export these stats to Presto coordinator. If not, do you plan to produce such a PR?

HolyLow commented:
@mbasmanova Not yet, I am willing to submit such a PR but recently I am working on other projects, so the delay caused by it might not be small. If you want the feature quickly, maybe you should produce the PR yourself. If you are not in a hurry, I could do it later.
